### PR TITLE
Blazor Web project interactive options improvements. Rename render modes.

### PR DESCRIPTION
**Pull request by SteveSandersonMS**
Fixes https://github.com/dotnet/aspnetcore/issues/50433 (Add root level interactivity option)
Fixes https://github.com/dotnet/aspnetcore/issues/50646 (Remove workaround for Counter component)
Fixes https://github.com/dotnet/aspnetcore/issues/50636 (Clarify the names of the interactive render modes)

In terms of the code we now emit, there should be nothing controversial here. The template just has to do quite a bit of if/else in many places to account for all these options and how rendermodes are used and not used based on them.

The PR is big because the renames have really wide impact, but almost all the "files changes" are just due to renames. The only real code changes are in the project templates.

# Testing impact

Adding this option, the BlazorWeb template now has **so many** possible combinations of options, including:

 - Whether or not to enable Server interactivity
 - Whether or not to enable WebAssembly interactivity
 - Whether or not to be interactive from the root
 - Whether or not to include sample content
 - Whether or not to use ProgramMain

So that's around 32 combinations of output - without even accounting for auth options! We don't currently have E2E verification of any of them, as those tests are skipped due to unreliability. We're going to have to lean hard on CTI validations for this, and make sure all the important combinations are covered - cc @mkArtakMSFT.

# Options design update

Having a list of 6 separate checkboxes in VS is pretty unpleasant and hard to understand:

So, in this PR I'm proposing (and have implemented, but we can still change it), a change to use dropdowns for the interactivity type and location options. This reduces the number of inputs by one, and means they can be more self-describing:

 * The "interactivity type" choices are:
   * **None**
   * **Server** (default)
   * **WebAssembly**
   * **Auto (Server and WebAssembly)**.
 * The "interactivity location" choices are:
   * **Per page/component** (default)
   * **Global**

Note that "interactivity location" is disabled if interactivity type == "None", but [only CLI honors that right now](https://github.com/dotnet/templating/issues/5648) (VS should add support later, and until then, location will have no effect if there's no interactivity).

I think this is much easier to understand, since you no longer have to infer that enabling both Server and WebAssembly means you're going to get Auto. It's also much better in the CLI, since it was completely ridiculous before that `--use-server` defaulted to true but `--use-wasm` defaulted to false, so to get WebAssembly you needed to set `--use-server false --use wasm`. Now you would say `--interactivity webassembly` (and not `wasm` - that was weird too).

image

### CODE REVIEW

**Comment by SteveSandersonMS** [comment]
image

**Review by danroth27 — APPROVED** [review]
Love the new UX! It's way better than the flat list of check boxes.

File: src/ProjectTemplates/test/Templates.Tests/template-baselines.json [hunk 1]
```diff
@@ -660,8 +748,8 @@
       "AuthOption": "None"
     },
     "EmptyUseWebAssembly": {
-      "Template": "web",
-      "Arguments": "new blazor --use-server false --use-wasm --empty",
+      "Template": "blazor",
```
Comment (mkArtakMSFT) [inline src/ProjectTemplates/test/Templates.Tests/template-baselines.json:751]: So this was a bug we didn't notice in the past then?

**Review by mkArtakMSFT — APPROVED** [review]

**Comment by davidfowl** [comment]
These names don't look very nice... I don't have an alternative yet but these feel very verbose and not that much more clear.

File: src/ProjectTemplates/test/Templates.Tests/template-baselines.json (hunk 1 above)
Comment (SteveSandersonMS) [reply src/ProjectTemplates/test/Templates.Tests/template-baselines.json:751]: It was a weird inconsistency but didn't impact the test actually working correctly. I think the `Template` string is only used for reporting in logs, and doesn't affect whether the test passes or fails.

**Comment by SteveSandersonMS** [comment]
Update: we can't use the parameter name `--interactive` because that's a [reserved alias](https://github.com/dotnet/templating/blob/1613af0baef18703492d66f82f7aacd2291fddce/docs/Reserved-Aliases.md?plain=1#L175). I've set it back to `--interactivity`.

**Pull request merged**